test/model/typing_status_test.dart

```diff
@@ -0,0 +1,240 @@
+import 'dart:async';
+
+import 'package:checks/checks.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:zulip/api/model/events.dart';
+import 'package:zulip/api/model/model.dart';
+import 'package:zulip/model/narrow.dart';
+import 'package:zulip/model/typing_status.dart';
+
+import '../example_data.dart' as eg;
+import '../fake_async.dart';
+
+extension TypingStatusChecks on Subject<TypingStatus> {
+  Subject<Map<SendableNarrow, Map<int, Timer>>> get typistsByNarrow => has((x) => x.typistIdsByNarrow, 'typistsByNarrow');
+}
+
+extension TimerChecks on Subject<Timer> {
+  Subject<bool> get isActive => has((x) => x.isActive, 'isActive');
+}
+
+void main() {
+  late TypingStatus model;
+  late int notifiedCount;
+
+  void prepareModel() {
+    model = TypingStatus(
+      selfUserId: eg.selfUser.userId,
+      typingStartedExpiryPeriod: const Duration(milliseconds: 15000));
+    check(model).typistsByNarrow.isEmpty();
+    notifiedCount = 0;
+    model.addListener(() => notifiedCount += 1);
+  }
+
+  void checkNotNotified() {
+    check(notifiedCount).equals(0);
+  }
+
+  void checkNotifiedOnce() {
+    check(notifiedCount).equals(1);
+    notifiedCount = 0;
+  }
+
+  Subject<TypingStatus> checkTyping(SendableNarrow narrow, TypingOp op, User typist) {
+    model.handleTypingEvent(eg.typingEvent(narrow, op, typist.userId));
+    return check(model);
+  }
+
+  final stream = eg.stream();
+  final topicNarrow = TopicNarrow(stream.streamId, 'foo');
+
+  final dmNarrow = DmNarrow.withUser(eg.otherUser.userId, selfUserId: eg.selfUser.userId);
+  final groupNarrow = DmNarrow.withOtherUsers(
+    [eg.otherUser.userId, eg.thirdUser.userId], selfUserId: eg.selfUser.userId);
+
+  anyTimer(Subject<Object?> it) => it.isA<Timer>();
+
+  group('handle typing start events', () {
+    test('add typists in separate narrows', () {
+      prepareModel();
+
+      checkTyping(dmNarrow, TypingOp.start, eg.otherUser)
+        .typistsByNarrow.deepEquals(
+          {dmNarrow: {eg.otherUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      checkTyping(groupNarrow, TypingOp.start, eg.thirdUser)
+        .typistsByNarrow.deepEquals({
+          dmNarrow: {eg.otherUser.userId: anyTimer},
+          groupNarrow: {eg.thirdUser.userId: anyTimer}});
```

Comment:
This is a bit confusing to read because the `checkTyping` function turns out to have a side effect, mutating the model, so e.g. the behavior at this point would be different if the other `checkTyping` call above were left out.

Better to make the side effects explicit, by separating the `model.handleTypingEvent` call from the `check(model)`:

```suggestion
      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.thirdUser.userId));
      check(model).typistsByNarrow.deepEquals({
        dmNarrow: {eg.otherUser.userId: anyTimer},
        groupNarrow: {eg.thirdUser.userId: anyTimer}});
```

Then that first line could still benefit from a helper function to reduce the boilerplate; but it can be focused just on that step, and as a result also it can have a name to match, without "check".